File: src/main/java/at/bitfire/vcard4android/BatchOperation.java
/*
 * Copyright (c) 2013 – 2015 Ricki Hirner (bitfire web engineering).
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 * PURPOSE.  See the GNU General Public License for more details.
 */

package at.bitfire.vcard4android;

import android.content.ContentProviderClient;
import android.content.ContentProviderOperation;
import android.content.ContentProviderResult;
import android.content.ContentUris;
import android.content.OperationApplicationException;
import android.os.RemoteException;
import android.os.TransactionTooLargeException;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

import lombok.NonNull;

public class BatchOperation {

    private final ContentProviderClient providerClient;
    private final List<Operation> queue = new LinkedList<>();
    private ContentProviderResult[] results;


    public BatchOperation(@NonNull ContentProviderClient providerClient) {
        this.providerClient = providerClient;
    }

    public int nextBackrefIdx() {
        return queue.size();
    }

    public void enqueue(Operation operation) {
        queue.add(operation);
    }

    public int commit() throws ContactsStorageException {
        int affected = 0;
        if (!queue.isEmpty())
            try {
                Constants.log.fine("Committing " + queue.size() + " operations …");

                results = new ContentProviderResult[queue.size()];
                runBatch(0, queue.size());

                for (ContentProviderResult result : results)
                    if (result != null)                 // will have either .uri or .count set
                        if (result.count != null)
                            affected += result.count;
                        else if (result.uri != null)
                            affected += 1;
                Constants.log.fine("… " + affected + " record(s) affected");

            } catch(OperationApplicationException|RemoteException e) {
                throw new ContactsStorageException("Couldn't apply batch operation", e);
            }

        queue.clear();

        return affected;
    }

    public ContentProviderResult getResult(int idx) {
        return results[idx];
    }


    /**
     * Runs a subset of the operations in {@link #queue} using {@link #providerClient} in a transaction.
     * Catches {@link TransactionTooLargeException} and splits the operations accordingly.
     * @param start    index of first operation which will be run (inclusive)
     * @param end      index of last operation which will be run (exclusive!)
     * @throws RemoteException  if the provider clients throws a {@link RemoteException}, or
     *                          if the transaction is too large and can't be split
     */
    private void runBatch(int start, int end) throws RemoteException, OperationApplicationException, ContactsStorageException {
        if (end == start)
            return;     // nothing to do

        try {
            Constants.log.fine("Running operations " + start + " to " + (end - 1));
            ContentProviderResult partResults[] = providerClient.applyBatch(toCPO(start, end));

            int n = end - start;
            if (partResults.length != n)
                throw new ContactsStorageException("Batch operation failed partially (only " + partResults.length + " of " + n + " operations done)");

            System.arraycopy(partResults, 0, results, start, n);
        } catch(TransactionTooLargeException e) {
            if (end <= start + 1)
                // only one operation, can't be split
                throw new RemoteException("Can't transfer data to content provider (data row too large)");

            Constants.log.warning("Transaction too large, splitting (losing atomicity)");
            int mid = start + (end - start)/2;
            runBatch(start, mid);
            runBatch(mid, end);
        }
    }

    private ArrayList<ContentProviderOperation> toCPO(int start, int end) {
        ArrayList<ContentProviderOperation> cpo = new ArrayList<>(end - start);

        int i = 0;
        for (Operation op : queue.subList(start, end)) {
            ContentProviderOperation.Builder builder = op.builder;

            // fill in back references
            if (op.backrefKey != null) {
                if (op.backrefIdx < start) {
                    // back reference is outside of the current batch
                    builder .withValueBackReferences(null)
                            .withValue(op.backrefKey, ContentUris.parseId(results[op.backrefIdx].uri));
                } else
                    // back reference is in current batch, apply offset
                    builder.withValueBackReference(op.backrefKey, op.backrefIdx - start);
            }

            // set a yield point at least every 450 operations
            if (i++ % 450 == 0)
                builder.withYieldAllowed(true);

            cpo.add(builder.build());
        }
        return cpo;
    }


    public static class Operation {

        final ContentProviderOperation.Builder builder;
        final String backrefKey;
        final int backrefIdx;

        public Operation(ContentProviderOperation.Builder builder) {
            this.builder = builder;
            backrefKey = null;
            backrefIdx = -1;
        }

        public Operation(ContentProviderOperation.Builder builder, String backrefKey, int backrefIdx) {
            this.builder = builder;
            this.backrefKey = backrefKey;
            this.backrefIdx = backrefIdx;
        }

    }

}

File: src/main/java/at/bitfire/vcard4android/BatchOperation.kt
/*
 * Copyright © Ricki Hirner (bitfire web engineering).
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Public License v3.0
 * which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/gpl.html
 */

package at.bitfire.vcard4android

import android.content.*
import android.net.Uri
import android.os.RemoteException
import android.os.TransactionTooLargeException
import java.util.*

class BatchOperation(
        private val providerClient: ContentProviderClient
) {

    companion object {

        /**
         * See https://android.googlesource.com/platform/packages/providers/ContactsProvider.git/+/refs/heads/android11-release/src/com/android/providers/contacts/AbstractContactsProvider.java#70
         *
         * Some operations may count more than one operation, so use a safe value of 450 instead of 500.
         */
        const val MAX_OPERATIONS_PER_YIELD_POINT = 450

    }

    private val queue = LinkedList<CpoBuilder>()
    private var results = arrayOfNulls<ContentProviderResult?>(0)


    fun nextBackrefIdx() = queue.size

    fun enqueue(operation: CpoBuilder): BatchOperation {
        queue.add(operation)
        return this
    }

    fun commit(): Int {
        var affected = 0
        if (!queue.isEmpty())
            try {
                Constants.log.fine("Committing ${queue.size} operations")

                results = arrayOfNulls(queue.size)
                runBatch(0, queue.size)

                for (result in results.filterNotNull())
                    when {
                        result.count != null -> affected += result.count ?: 0
                        result.uri != null   -> affected += 1
                    }
                Constants.log.fine("… $affected record(s) affected")

            } catch(e: Exception) {
                throw ContactsStorageException("Couldn't apply batch operation", e)
            }

        queue.clear()
        return affected
    }

    fun getResult(idx: Int) = results[idx]


    /**
     * Runs a subset of the operations in [queue] using [providerClient] in a transaction.
     * Catches [TransactionTooLargeException] and splits the operations accordingly.
     * @param start index of first operation which will be run (inclusive)
     * @param end   index of last operation which will be run (exclusive!)
     * @throws RemoteException on calendar provider errors
     * @throws OperationApplicationException when the batch can't be processed
     * @throws CalendarStorageException if the transaction is too large
     */
    private fun runBatch(start: Int, end: Int) {
        if (end == start)
            return     // nothing to do

        try {
            val ops = toCPO(start, end)
            Constants.log.fine("Running ${ops.size} operations ($start .. ${end-1})")
            val partResults = providerClient.applyBatch(ops)

            val n = end - start
            if (partResults.size != n)
                Constants.log.warning("Batch operation returned only ${partResults.size} instead of $n results")

            System.arraycopy(partResults, 0, results, start, partResults.size)
        } catch(e: TransactionTooLargeException) {
            if (end <= start + 1)
                // only one operation, can't be split
                throw ContactsStorageException("Can't transfer data to content provider (data row too large)")

            Constants.log.warning("Transaction too large, splitting (losing atomicity)")
            val mid = start + (end - start)/2
            runBatch(start, mid)
            runBatch(mid, end)
        }
    }

    private fun toCPO(start: Int, end: Int): ArrayList<ContentProviderOperation> {
        val cpo = ArrayList<ContentProviderOperation>(end - start)

        for ((i, cpoBuilder) in queue.subList(start, end).withIndex()) {
            for ((backrefKey, backrefIdx) in cpoBuilder.valueBackrefs) {
                if (backrefIdx < start) {
                    // back reference is outside of the current batch, get result from previous execution ...
                    val resultUri = results[backrefIdx]?.uri ?: throw ContactsStorageException("Referenced operation didn't produce a valid result")
                    val resultId = ContentUris.parseId(resultUri)
                    // ... and use result directly instead of using a back reference
                    cpoBuilder  .removeValueBackReference(backrefKey)
                                .withValue(backrefKey, resultId)
                } else
                    // back reference is in current batch, shift index
                    cpoBuilder.withValueBackReference(backrefKey, backrefIdx - start)
            }

            if (i % MAX_OPERATIONS_PER_YIELD_POINT == MAX_OPERATIONS_PER_YIELD_POINT - 1)
                cpoBuilder.yieldAllowed = true

            cpo += cpoBuilder.build()
        }
        return cpo
    }


    /**
     * Wrapper for [ContentProviderOperation.Builder] that allows to reset previously-set
     * value back references.
     */
    class CpoBuilder private constructor(
            val uri: Uri,
            val type: Type
    ) {

        enum class Type { INSERT, UPDATE, DELETE }

        companion object {

            fun newInsert(uri: Uri) = CpoBuilder(uri, Type.INSERT)
            fun newUpdate(uri: Uri) = CpoBuilder(uri, Type.UPDATE)
            fun newDelete(uri: Uri) = CpoBuilder(uri, Type.DELETE)

        }


        var selection: String? = null
        var selectionArguments: Array<String>? = null

        val values = mutableMapOf<String, Any?>()
        val valueBackrefs = mutableMapOf<String, Int>()

        var yieldAllowed = false


        fun withSelection(select: String, args: Array<String>): CpoBuilder {
            selection = select
            selectionArguments = args
            return this
        }

        fun withValueBackReference(key: String, index: Int): CpoBuilder {
            valueBackrefs[key] = index
            return this
        }

        fun removeValueBackReference(key: String): CpoBuilder {
            if (valueBackrefs.remove(key) == null)
                throw IllegalArgumentException("$key was not set as value back reference")
            return this
        }

        fun withValue(key: String, value: Any?): CpoBuilder {
            values[key] = value
            return this
        }


        fun build(): ContentProviderOperation {
            val builder = when (type) {
                Type.INSERT -> ContentProviderOperation.newInsert(uri)
                Type.UPDATE -> ContentProviderOperation.newUpdate(uri)
                Type.DELETE -> ContentProviderOperation.newDelete(uri)
            }

            if (selection != null)
                builder.withSelection(selection, selectionArguments)

            for ((key, value) in values)
                builder.withValue(key, value)
            for ((key, index) in valueBackrefs)
                builder.withValueBackReference(key, index)

            if (yieldAllowed)
                builder.withYieldAllowed(true)

            return builder.build()
        }

    }

}